File: site/src/lib/entity-families.ts
/**
 * entity-families.ts — merge the curated corporate families into the
 * `/companies/` table (PM Sprint 2, §P1-3). PURE (no fs, no React).
 *
 * The defect: `/companies/` listed **RAYTHEON COMPANY $43.7B (#4)** and
 * **RTX CORP $24.6B (#6)** as two families. One company, renamed in 2023 —
 * the split understated the combined position by roughly half and misordered
 * the top ten.
 *
 * The merge, and its guard rails:
 *
 *  - Membership comes ONLY from the curated seed (data-seeds/
 *    entity_family_events.csv → entity_family_events.json). Nothing here
 *    infers a family from a name.
 *  - A registry row belongs to AT MOST ONE family (the exporter raises if the
 *    seed ever says otherwise), so `mergeCompanies` cannot emit a row twice.
 *    `assertNoDoubleCount` re-checks that on the OUTPUT: every input row
 *    appears in exactly one output row, and the merged obligation equals the
 *    payload's cited combined figure.
 *  - The combined figure is NOT summed in the browser. It is the exporter's
 *    derived `combined_obligation_fact_id` — a citation whose inputs are the
 *    member facts and whose sum the citation verifier recomputes. The site
 *    renders a cited number, never an arithmetic one.
 *  - Members outside the top-200 export contribute to the family's total and
 *    are named, but carry no `/company/` link (they have no page).
 *
 * Ranking is recomputed after merging, so the top ten is correct.
 */

import type { EntityTop } from "@/lib/data";

// ── Payload (data/site/json/entity_family_events.json) ──────────────────────

export type FamilyEventKind = "rename" | "acquisition" | "merger";

/** How well the linked source supports THIS row (exporter EVIDENCE_KINDS). */
export type FamilyEvidence = "sourced" | "name-inferred";

export interface FamilyEventRow {
  from_name: string;
  to_name: string;
  event: FamilyEventKind;
  effective_date: string;
  evidence: FamilyEvidence;
  /** Official press release or SEC filing — an EXTERNAL reference. */
  source_url: string;
  /** Form label for the link ("8-K Item 2.01", "Press release"). */
  source_form: string;
  /** The document's own date. */
  source_date: string;
  /** When a curator last opened the source and checked this row. */
  source_verified: string;
  note: string;
  from_family_key: string | null;
  to_family_key: string | null;
  /** DOM id of this event's row on /companies/families/. */
  anchor: string;
  /** Registry family keys whose membership THIS event explains (may be []). */
  changed_family_keys: string[];
}

/**
 * The event that explains why ONE registry name is in the family (§P1-3 fix
 * round). Computed by the exporter — see entity_families.member_arrivals.
 *
 * The defect it closes: /companies/ hung ONE trailing event label off a
 * heterogeneous list of former names, so EXELIS INC. (acquired by Harris in
 * 2015) read "acquired 2019" — the date of the separate Harris/L3 merger —
 * and ROCKWELL COLLINS, INC. (acquired 2018) read "renamed 2023".
 */
export interface MemberArrival {
  event_index: number;
  /** /companies/families/#{anchor} — the exact row that documents this. */
  anchor: string;
  /** "from": this name is what changed. "to": this is the post-event name. */
  role: "from" | "to";
  /** The other side of the event. */
  counterparty: string;
  event: FamilyEventKind;
  effective_date: string;
  evidence: FamilyEvidence;
}

export interface FamilyMember {
  family_key: string;
  display_name: string;
  slug: string;
  /** True when this member is in the top-200 export (has a /company/ page). */
  has_page: boolean;
  total_obligation: number;
  total_obligation_fact_id: string | null;
  uei_count: number;
  worst_confidence: string;
  /** null for the family's surviving name (the acquirer is not "acquired"). */
  arrival: MemberArrival | null;
}

export interface CuratedFamily {
  label: string;
  slug: string;
  members: FamilyMember[];
  /** null when the family has fewer than two resolved members (no merge). */
  combined_obligation: number | null;
  combined_obligation_fact_id: string | null;
  combined_uei_count: number;
  events: FamilyEventRow[];
}

export interface FamilyEventsPayload {
  schema_version: number;
  method: string;
  source_kind: string;
  seed: string;
  /** Max source_verified across the seed — the table's "as of" stamp. */
  sources_verified_through: string | null;
  families: CuratedFamily[];
}

// ── Merged rows ─────────────────────────────────────────────────────────────

/**
 * One `/companies/` row. A plain registry row and a merged corporate family
 * render through the SAME shape, so the table has one code path.
 */
export interface CompanyRow {
  /** Family slug for a merge, entity slug for a plain row (React key). */
  key: string;
  /** The name in the first column. */
  displayName: string;
  /** Registry rows folded into this line (1 for an unmerged row). */
  members: FamilyMember[];
  totalObligation: number;
  /** The CITED fact for `totalObligation` — combined fact for a merge. */
  factId: string | null;
  ueiCount: number;
  worstConfidence: string;
  /** Curated family this row came from — null for a plain registry row. */
  family: CuratedFamily | null;
  /** True when ≥2 registry rows were folded together. */
  merged: boolean;
  /** 1-based rank by obligation AFTER the merge. */
  rank: number;
}

/** Confidence ordering, worst first — a family is as good as its worst member. */
const CONFIDENCE_RANK: Record<string, number> = { low: 0, medium: 1, high: 2 };

export function worstConfidence(values: readonly string[]): string {
  let worst = values[0] ?? "medium";
  for (const v of values) {
    if ((CONFIDENCE_RANK[v] ?? 1) < (CONFIDENCE_RANK[worst] ?? 1)) worst = v;
  }
  return worst;
}

function entityToMember(e: EntityTop): FamilyMember {
  return {
    family_key: e.family_key,
    display_name: e.display_name,
    slug: e.slug,
    has_page: true,
    total_obligation: e.total_obligation,
    total_obligation_fact_id: e.total_obligation_fact_id,
    uei_count: e.uei_count,
    worst_confidence: e.worst_confidence,
    arrival: null, // an unmerged registry row arrived by no curated event
  };
}

// ── Event vocabulary ────────────────────────────────────────────────────────

/**
 * The past-participle label for an event kind. "merger" exists because calling
 * the UTC/Raytheon merger of equals an "acquisition" misstates who absorbed
 * whom, and calling it a "rename" is worse.
 */
const EVENT_VERB: Record<FamilyEventKind, string> = {
  rename: "renamed",
  acquisition: "acquired",
  merger: "merged",
};

/** "acquired 2015" / "renamed 2023" / "merged 2020" — the year is the EVENT's. */
export function eventLabel(event: FamilyEventKind, effectiveDate: string): string {
  return `${EVENT_VERB[event] ?? event} ${effectiveDate.slice(0, 4)}`;
}

/**
 * The per-former-name annotation. A "from" arrival states what happened to
 * THAT name; a "to" arrival names the predecessor it replaced, because
 * "Northrop Grumman Innovation Systems — acquired 2018" alone would leave the
 * reader wondering which company that was.
 */
export function memberEventLabel(arrival: MemberArrival): string {
  const label = eventLabel(arrival.event, arrival.effective_date);
  return arrival.role === "to"
    ? `formerly ${arrival.counterparty}, ${label}`
    : label;
}

/**
 * Merge the top-200 registry rows into corporate families and re-rank.
 *
 * A family is merged only when it has a cited combined figure — i.e. the
 * exporter minted `combined_obligation_fact_id`. A family whose combined fact
 * is missing (fewer than two resolved members, or an unresolvable citation)
 * leaves its rows alone: the site would rather show the honest split than an
 * uncited merged number.
 */
export function mergeCompanies(
  companies: readonly EntityTop[],
  payload: FamilyEventsPayload | null,
): CompanyRow[] {
  const families = (payload?.families ?? []).filter(
    (f) => f.combined_obligation != null && f.combined_obligation_fact_id != null,
  );

  // family_key → the family that owns it (at most one, exporter-enforced).
  const owner = new Map<string, CuratedFamily>();
  for (const fam of families) {
    for (const m of fam.members) owner.set(m.family_key, fam);
  }

  const rows: CompanyRow[] = [];
  const emitted = new Set<string>();

  for (const company of companies) {
    const fam = owner.get(company.family_key);
    if (!fam) {
      rows.push({
        key: company.slug,
        displayName: company.display_name,
        members: [entityToMember(company)],
        totalObligation: company.total_obligation,
        factId: company.total_obligation_fact_id,
        ueiCount: company.uei_count,
        worstConfidence: company.worst_confidence,
        family: null,
        merged: false,
        rank: 0,
      });
      continue;
    }
    if (emitted.has(fam.slug)) continue; // already folded in at its first member
    emitted.add(fam.slug);
    rows.push({
      key: fam.slug,
      displayName: fam.label,
      members: fam.members,
      totalObligation: fam.combined_obligation!,
      factId: fam.combined_obligation_fact_id,
      ueiCount: fam.combined_uei_count,
      worstConfidence: worstConfidence(fam.members.map((m) => m.worst_confidence)),
      family: fam,
      merged: fam.members.length > 1,
      rank: 0,
    });
  }

  rows.sort((a, b) => b.totalObligation - a.totalObligation);
  rows.forEach((r, i) => {
    r.rank = i + 1;
  });
  return rows;
}

/**
 * The double-count guard, asserted on the OUTPUT.
 *
 * Returns the list of violations (empty when the merge held):
 *   - a registry family_key appearing in more than one rendered row;
 *   - a merged row whose member totals do not sum to the cited combined
 *     figure (a member shown but not counted, or counted twice);
 *   - an input company that vanished entirely.
 *
 * Called by the page at build time (throws — a build failure, never a silent
 * bad number) and independently by gate 24 leg (g) on the built HTML.
 */
export function assertNoDoubleCount(
  companies: readonly EntityTop[],
  rows: readonly CompanyRow[],
): string[] {
  const problems: string[] = [];

  const seen = new Map<string, string>();
  for (const row of rows) {
    for (const m of row.members) {
      const prev = seen.get(m.family_key);
      if (prev && prev !== row.key) {
        problems.push(
          `registry family ${m.family_key} appears in two rows (${prev}, ${row.key}) — its obligations would be counted twice`,
        );
      }
      seen.set(m.family_key, row.key);
    }
  }

  for (const c of companies) {
    if (!seen.has(c.family_key)) {
      problems.push(`registry family ${c.family_key} vanished from the merged table`);
    }
  }

  for (const row of rows) {
    if (!row.merged) continue;
    const sum = row.members.reduce((n, m) => n + m.total_obligation, 0);
    // Float sum of ~1e11 magnitudes — a cent of tolerance.
    if (Math.abs(sum - row.totalObligation) > 0.01) {
      problems.push(
        `merged row ${row.key}: members sum to ${sum} but the cited combined figure is ${row.totalObligation}`,
      );
    }
    const keys = new Set(row.members.map((m) => m.family_key));
    if (keys.size !== row.members.length) {
      problems.push(`merged row ${row.key} lists a member twice`);
    }
  }

  return problems;
}

/**
 * Should the per-row confidence chip render?
 *
 * §P1-3: all 200 rows read "medium" — a chip that never varies conveys
 * nothing while adding 200 amber badges of visual alarm. Suppress it when it
 * is uniform across the WHOLE table (not the filtered view — a chip column
 * that appears and disappears as the reader types would be worse), and state
 * the method once in the header instead.
 */
export function confidenceIsUniform(rows: readonly CompanyRow[]): boolean {
  if (rows.length === 0) return true;
  const first = rows[0].worstConfidence;
  return rows.every((r) => r.worstConfidence === first);
}

/** Every fact_id a merged table needs in its citation slice. */
export function companyRowFactIds(rows: readonly CompanyRow[]): string[] {
  const ids: string[] = [];
  for (const row of rows) {
    if (row.factId) ids.push(row.factId);
    for (const m of row.members) {
      if (m.total_obligation_fact_id) ids.push(m.total_obligation_fact_id);
    }
  }
  return ids;
}

/** All curated events, newest first — the events page's table order. */
export function allEvents(
  payload: FamilyEventsPayload | null,
): { family: CuratedFamily; event: FamilyEventRow }[] {
  const out: { family: CuratedFamily; event: FamilyEventRow }[] = [];
  for (const family of payload?.families ?? []) {
    for (const event of family.events) out.push({ family, event });
  }
  out.sort((a, b) => {
    if (a.event.effective_date !== b.event.effective_date) {
      return a.event.effective_date < b.event.effective_date ? 1 : -1;
    }
    return a.family.label.localeCompare(b.family.label);
  });
  return out;
}

File: src/govbudget/entity_families.py
"""Curated corporate rename/acquisition events (PM Sprint 2, spec §P1-3).

`/companies/` listed **RAYTHEON COMPANY $43.7B (#4)** and **RTX CORP $24.6B (#6)**
as two families. They are one company — Raytheon renamed to RTX in 2023 — and
the split understated the combined position by roughly half and misordered the
top ten. Entity resolution cannot see this: the warehouse's `family_key` is a
name-inference over award recipients, and a rename produces two names.

This module carries the fix that name inference cannot: a HAND-CURATED table of
well-documented corporate events, each with an OFFICIAL source (SEC filing or
company press release). It is deliberately NOT a warehouse citation tier —
these rows cite documents outside the lake, they are rendered as explicit
external references, and the site says so plainly. Nothing here is inferred.

Contract:
  * Source of truth: ``data-seeds/entity_family_events.csv`` with columns
    ``family, from_name, to_name, event, effective_date, evidence, source_url,
    source_form, source_date, source_verified, note``.
  * Endpoints resolve to warehouse ``family_key`` values through the SAME
    normalizer that MINTED those keys (``govbudget.entities.normalize_name``)
    and only on EXACT normalized equality — no fuzzy matching, ever. An
    endpoint that does not resolve is recorded as unresolved and rendered as
    such; it is never guessed at.

    CURATOR HAZARD, learned the hard way in the 2026-08-04 source audit: exact
    normalized equality can still be a FALSE positive, because the normalizer
    strips corporate suffixes. "United Technologies Corporation" normalizes to
    ``UNITED TECHNOLOGIES``, which is a real warehouse family — an unrelated
    USD 1.5M "UNITED TECHNOLOGIES, LLC", not the aerospace parent. Writing
    that endpoint would have merged a stranger into RTX. When an endpoint is a
    corporate parent whose award-data family is NOT the same legal entity,
    write the name the SOURCE uses for the business ("Collins Aerospace
    Systems (United Technologies Corporation)") and check the resulting key
    against ``dim_entities.family_key`` before committing.
  * A ``family_key`` may belong to AT MOST ONE curated family. Two families
    claiming the same key is the double-count hazard in its seed form, so it
    is a load-time error, not a runtime surprise.
  * Members are DISJOINT by warehouse construction — ``entity_xwalk`` assigns
    each ``recipient_uei`` exactly one ``family_key`` — so the combined total
    is the plain sum of member totals with no risk of double counting. The
    exporter re-asserts that disjointness against the lake at build time.

Loud failures only: a malformed row, an unknown event kind, a non-https source,
or a cross-family key collision raises. A quietly-dropped curated event would
silently restore the very defect this table exists to close.
"""

from __future__ import annotations

import csv
import re
from dataclasses import dataclass, field
from datetime import date
from pathlib import Path

from govbudget.entities import normalize_name

# The seed's exact column set (locked in the Sprint 2 plan; widened by the
# Sprint 2 visual-judge fix round with the chain-of-custody + evidence fields).
REQUIRED_COLUMNS = (
    "family",
    "from_name",
    "to_name",
    "event",
    "effective_date",
    "evidence",
    "source_url",
    "source_form",
    "source_date",
    "source_verified",
    "note",
)

# "merger" was added in the fix round: calling the UTC/Raytheon merger of
# equals — or the HPE-ES/CSC spin-merge that CREATED DXC — an "acquisition"
# misstates who absorbed whom, and calling it a "rename" is worse. The kind a
# row carries is the kind ITS OWN SOURCE uses.
EVENT_KINDS = ("rename", "acquisition", "merger")

#: How well the SOURCE supports this exact row.
#:   sourced       — the linked document states this event for THESE names.
#:   name-inferred — the document states the corporate event, but not for this
#:                   specific award-data recipient; the link from the recipient
#:                   name to the filing party is OURS. Rendered as such.
EVIDENCE_KINDS = ("sourced", "name-inferred")

_ISO_DATE = re.compile(r"^\d{4}-\d{2}-\d{2}$")
_SLUG_STRIP = re.compile(r"[^a-z0-9]+")


class FamilyEventsError(ValueError):
    """Raised on any malformed or self-contradicting curated row."""


def slugify(label: str) -> str:
    """URL slug for a family label ("L3Harris Technologies" → l3harris-technologies")."""
    return _SLUG_STRIP.sub("-", label.strip().lower()).strip("-")


@dataclass(frozen=True)
class FamilyEvent:
    """One curated corporate event, exactly as authored in the seed."""

    family: str
    from_name: str
    to_name: str
    event: str
    effective_date: str
    #: EVIDENCE_KINDS — how well the source supports THIS row.
    evidence: str
    source_url: str
    #: Human form label for the link ("8-K Item 2.01", "Press release").
    source_form: str
    #: The document's own date (SEC filing date / press-release date).
    source_date: str
    #: The date a curator last opened the source and checked this row.
    source_verified: str
    note: str


def load_family_events(csv_path: Path) -> list[FamilyEvent]:
    """Parse + validate the curated seed. Raises FamilyEventsError on any defect."""
    if not csv_path.exists():
        raise FamilyEventsError(
            f"curated entity-family seed missing: {csv_path} — /companies/ would "
            f"silently go back to splitting renamed companies"
        )
    with csv_path.open(newline="", encoding="utf-8") as fh:
        reader = csv.DictReader(fh)
        header = tuple(reader.fieldnames or ())
        if header != REQUIRED_COLUMNS:
            raise FamilyEventsError(
                f"{csv_path.name}: columns {header} != required {REQUIRED_COLUMNS}"
            )
        events: list[FamilyEvent] = []
        for i, row in enumerate(reader, start=2):  # line 1 is the header
            def need(col: str) -> str:
                value = (row.get(col) or "").strip()
                if not value:
                    raise FamilyEventsError(f"{csv_path.name}:{i}: empty {col}")
                return value

            event = need("event")
            if event not in EVENT_KINDS:
                raise FamilyEventsError(
                    f"{csv_path.name}:{i}: event {event!r} not in {EVENT_KINDS}"
                )
            evidence = need("evidence")
            if evidence not in EVIDENCE_KINDS:
                raise FamilyEventsError(
                    f"{csv_path.name}:{i}: evidence {evidence!r} not in "
                    f"{EVIDENCE_KINDS} — a row must declare whether its source "
                    f"names these exact parties or whether we inferred the link"
                )

            def need_date(col: str) -> str:
                value = need(col)
                if not _ISO_DATE.match(value):
                    raise FamilyEventsError(
                        f"{csv_path.name}:{i}: {col} {value!r} is not YYYY-MM-DD"
                    )
                try:
                    date.fromisoformat(value)
                except ValueError as exc:
                    raise FamilyEventsError(
                        f"{csv_path.name}:{i}: {col} {value!r}: {exc}"
                    ) from exc
                return value

            effective_date = need_date("effective_date")
            source_date = need_date("source_date")
            source_verified = need_date("source_verified")
            if source_verified < source_date:
                raise FamilyEventsError(
                    f"{csv_path.name}:{i}: source_verified {source_verified!r} "
                    f"precedes source_date {source_date!r} — a row cannot have "
                    f"been checked before its source existed"
                )
            source_url = need("source_url")
            if not source_url.startswith("https://"):
                raise FamilyEventsError(
                    f"{csv_path.name}:{i}: source_url must be https (got {source_url!r}) "
                    f"— these render as external references, not warehouse citations"
                )
            from_name = need("from_name")
            to_name = need("to_name")
            if normalize_name(from_name) == normalize_name(to_name):
                raise FamilyEventsError(
                    f"{csv_path.name}:{i}: from_name and to_name normalize identically "
                    f"({normalize_name(from_name)!r}) — that records no event"
                )
            events.append(
                FamilyEvent(
                    family=need("family"),
                    from_name=from_name,
                    to_name=to_name,
                    event=event,
                    effective_date=effective_date,
                    evidence=evidence,
                    source_url=source_url,
                    source_form=need("source_form"),
                    source_date=source_date,
                    source_verified=source_verified,
                    note=need("note"),
                )
            )
    if not events:
        raise FamilyEventsError(f"{csv_path.name}: no rows")
    return events


@dataclass(frozen=True)
class ResolvedEndpoint:
    """One side of an event, and the warehouse family it resolved to (or not)."""

    name: str
    family_key: str | None

    @property
    def resolved(self) -> bool:
        return self.family_key is not None


@dataclass
class ResolvedEvent:
    event: FamilyEvent
    from_endpoint: ResolvedEndpoint
    to_endpoint: ResolvedEndpoint


@dataclass
class ResolvedFamily:
    """A curated family: its label, its warehouse members, and its events."""

    label: str
    slug: str
    #: warehouse family_keys, deduped; order is decided by the caller (by size).
    member_keys: list[str] = field(default_factory=list)
    events: list[ResolvedEvent] = field(default_factory=list)


def resolve_families(
    events: list[FamilyEvent],
    known_family_keys: set[str],
) -> list[ResolvedFamily]:
    """Group curated events into families and resolve their endpoints.

    ``known_family_keys`` is the warehouse's ``dim_entities.family_key`` set.
    Resolution is exact normalized equality — the same normalizer that minted
    those keys. Endpoints that do not resolve are kept (and reported), never
    guessed at.

    Raises FamilyEventsError when two curated families claim the same
    ``family_key`` — the seed-level form of a double count.
    """
    by_label: dict[str, ResolvedFamily] = {}
    order: list[str] = []
    for ev in events:
        if ev.family not in by_label:
            by_label[ev.family] = ResolvedFamily(
                label=ev.family, slug=slugify(ev.family)
            )
            order.append(ev.family)
        fam = by_label[ev.family]

        def endpoint(name: str) -> ResolvedEndpoint:
            key = normalize_name(name)
            return ResolvedEndpoint(
                name=name, family_key=key if key in known_family_keys else None
            )

        resolved = ResolvedEvent(ev, endpoint(ev.from_name), endpoint(ev.to_name))
        fam.events.append(resolved)
        for ep in (resolved.from_endpoint, resolved.to_endpoint):
            if ep.family_key and ep.family_key not in fam.member_keys:
                fam.member_keys.append(ep.family_key)

    # The family LABEL itself is a member when it names a warehouse family
    # (e.g. "Northrop Grumman" → NORTHROP GRUMMAN): the anchor row must be
    # inside its own family, or the merge would leave the parent behind.
    for label in order:
        fam = by_label[label]
        anchor = normalize_name(label)
        if anchor in known_family_keys and anchor not in fam.member_keys:
            fam.member_keys.insert(0, anchor)

    # A family_key may belong to at most one curated family.
    owner: dict[str, str] = {}
    for label in order:
        for key in by_label[label].member_keys:
            if key in owner and owner[key] != label:
                raise FamilyEventsError(
                    f"family_key {key!r} is claimed by both {owner[key]!r} and "
                    f"{label!r} — a key in two families would double-count its "
                    f"obligations on /companies/"
                )
            owner[key] = label

    # Drop families with fewer than one resolved member: an all-unresolved
    # family merges nothing. Its EVENTS still ship (the table is the asset) —
    # the caller keeps the full event list separately.
    return [by_label[label] for label in order]


# ── Per-member arrival: WHICH event explains THIS registry name ──────────────
#
# The defect this closes (Sprint 2 visual-judge round, both judges): /companies/
# hung ONE trailing event label off a heterogeneous list of former names. The
# RTX line read "RAYTHEON COMPANY · RTX CORP · ROCKWELL COLLINS … — renamed
# 2023", and the L3Harris line labelled EXELIS INC. "acquired 2019". Exelis was
# acquired by Harris in **2015**; 2019 is the separate Harris/L3 merger. Two
# events, four years apart, collapsed into one wrong date — and a rename label
# pinned to companies that arrived by acquisition.
#
# The fix is in the MODEL, not the CSS: every former name carries its own
# event. This function computes that binding once, on the Python side, where
# `normalize_name` (which decides who the family's anchor is) already lives.


@dataclass(frozen=True)
class MemberArrival:
    """The single event that explains why a registry family is in this family."""

    #: Index of the event in ``ResolvedFamily.events`` — the /families/ anchor.
    event_index: int
    #: "from" — this name is what changed. "to" — this name is the post-event
    #: name of a predecessor the award data does not register separately.
    role: str
    #: The other side of the event, for a "formerly …" reading.
    counterparty: str


def member_arrivals(family: ResolvedFamily) -> dict[str, MemberArrival]:
    """Map each member ``family_key`` to the event that explains its membership.

    Three cases, in priority order:

    1. The member is an event's ``from`` endpoint — that event is what happened
       TO this name ("Exelis Inc., acquired 2015"). Always wins: it is the most
       specific true statement about the name the reader is looking at.
    2. The member is an event's ``to`` endpoint AND is not the family's own
       anchor — the member is the post-event name of a predecessor the award
       data does not register separately ("Northrop Grumman Innovation Systems,
       formerly Orbital ATK, acquired 2018").
    3. Otherwise — no arrival. This is the family's surviving name (Huntington
       Ingalls, SAIC, TransDigm are the *acquirers* in their rows; labelling
       them "acquired" would invert the transaction) or an anchor the seed
       pulled in by label.

    Deterministic under multiple candidates: earliest ``effective_date`` wins,
    then seed order.
    """
    anchor = normalize_name(family.label)
    from_hits: dict[str, list[tuple[str, int]]] = {}
    to_hits: dict[str, list[tuple[str, int]]] = {}
    for i, rev in enumerate(family.events):
        if rev.from_endpoint.family_key:
            from_hits.setdefault(rev.from_endpoint.family_key, []).append(
                (rev.event.effective_date, i)
            )
        if rev.to_endpoint.family_key:
            to_hits.setdefault(rev.to_endpoint.family_key, []).append(
                (rev.event.effective_date, i)
            )

    out: dict[str, MemberArrival] = {}
    for key in family.member_keys:
        if key in from_hits:
            _, idx = min(from_hits[key])
            out[key] = MemberArrival(
                event_index=idx,
                role="from",
                counterparty=family.events[idx].event.to_name,
            )
        elif key in to_hits and key != anchor:
            _, idx = min(to_hits[key])
            out[key] = MemberArrival(
                event_index=idx,
                role="to",
                counterparty=family.events[idx].event.from_name,
            )
    return out


def event_changed_keys(family: ResolvedFamily) -> list[list[str]]:
    """Per event (seed order): the member keys whose membership IT explains.

    The other half of the judges' finding: on /families/ every visible RTX
    event showed "no separate registry family" on one side, so *none of the
    events explained the merge it caused*. An event with an empty list here
    genuinely moved no registry row (Sikorsky), and the page now says so in
    those words instead of leaving the reader to infer it from two dashes.
    """
    arrivals = member_arrivals(family)
    out: list[list[str]] = [[] for _ in family.events]
    for key, arrival in arrivals.items():
        out[arrival.event_index].append(key)
    for keys in out:
        keys.sort()
    return out
